rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -1414,31 +1602,223 @@ void PlayerImpl::run_play_msg_post_callbacks(
+    message_published = true;
+  } catch (const std::exception & e) {
+    RCLCPP_ERROR_STREAM(
+      owner_->get_logger(), "Failed to send request on '" <<
+        rosbag2_cpp::service_event_topic_name_to_service_name(message->topic_name) <<
+        "' service. \nError: " << e.what());
+  }
+
+  // Calling on play message post-callbacks
+  run_play_msg_post_callbacks(message);
+  return message_published;
+}
+
+bool PlayerImpl::publish_message_by_play_action_client(
+  PlayerActionClientSharedPtr & action_client,
+  rosbag2_cpp::ActionInterfaceType action_interface_type,
+  rosbag2_storage::SerializedBagMessageSharedPtr & message)
+{
+  std::shared_ptr<uint8_t[]> type_erased_service_event;
+  PlayerActionClient::ServiceEventType service_event_type;
+  switch (action_interface_type) {
+    case rosbag2_cpp::ActionInterfaceType::SendGoalEvent:
+      type_erased_service_event =
+        action_client->deserialize_send_goal_service_event(*message->serialized_data);
+      service_event_type = action_client->get_service_event_type(
+        type_erased_service_event,
+        PlayerActionClient::ServiceInterfaceInAction::SEND_GOAL_SERVICE);
+      break;
+    case rosbag2_cpp::ActionInterfaceType::CancelGoalEvent:
+      type_erased_service_event =
+        action_client->deserialize_cancel_goal_service_event(*message->serialized_data);
+      service_event_type = action_client->get_service_event_type(
+        type_erased_service_event,
+        PlayerActionClient::ServiceInterfaceInAction::CANCEL_GOAL_SERVICE);
+      break;
+    case rosbag2_cpp::ActionInterfaceType::GetResultEvent:
+      type_erased_service_event =
+        action_client->deserialize_get_result_service_event(*message->serialized_data);
+      service_event_type = action_client->get_service_event_type(
+        type_erased_service_event,
+        PlayerActionClient::ServiceInterfaceInAction::GET_RESULT_SERVICE);
+      break;
+    default:
+      // Never go here
+      break;
+  }
+
+  if (!type_erased_service_event) {
+    return false;
+  }
+
+  // Ignore all response messages
+  if (service_event_type == PlayerActionClient::ServiceEventType::RESPONSE_SENT ||
+    service_event_type == PlayerActionClient::ServiceEventType::RESPONSE_RECEIVED)
+  {
+    return true;
+  }
+
+  // The request sent comes from the server or client side matches the user setting.
+  if (play_options_.service_requests_source == ServiceRequestsSource::SERVER_INTROSPECTION) {
+    if (service_event_type == PlayerActionClient::ServiceEventType::REQUEST_SENT) {
+      // Ignore request message from service client
+      return true;
+    }
+  } else {
+    // ServiceRequestsSource::CLIENT_INTROSPECTION
+    if (service_event_type == PlayerActionClient::ServiceEventType::REQUEST_RECEIVED) {
+      // Ignore request message from service server
+      return true;
+    }
+  }
+
+  try {
+    if (!action_client->generic_client()->wait_for_action_server(
```

Comment:
Can we avoid calling `wait_for_action_server(..)` inside the publishing method?
It could significantly delay other messages. We want to be close to the deterministic and reproducible behavior. 
This wait with a timeout will ruin reproducibility and determinism.